tests/integration/domains/osu_test.py

```diff
@@ -0,0 +1,208 @@
+        raise AssertionError(
+            "cho-token is not a valid uuid",
+            response.headers["cho-token"],
+        )
+
+    has_supporter = True
+
+    beatmap_md5 = "1cf5b2c2edfafd055536d2cefcb89c0e"
+    n300 = 83
+    n100 = 14
+    n50 = 5
+    ngeki = 23
+    nkatu = 6
+    nmiss = 6
+    score = 26810
+    max_combo = 52
+    perfect = False
+    grade = "C"
+    mods = 136
+    passed = True
+    game_mode = 0
+    client_time = datetime.now()
+
+    storyboard_md5 = hashlib.md5(b"lol123").hexdigest()
+
+    score_online_checksum = hashlib.md5(
+        "chickenmcnuggets{0}o15{1}{2}smustard{3}{4}uu{5}{6}{7}{8}{9}{10}{11}Q{12}{13}{15}{14:%y%m%d%H%M%S}{16}{17}".format(
+            n100 + n300,
+            n50,
+            ngeki,
+            nkatu,
+            nmiss,
+            beatmap_md5,
+            max_combo,
+            perfect,
+            username,
+            score,
+            grade,
+            mods,
+            passed,
+            game_mode,
+            client_time,
+            osu_version,
+            client_hashes,
+            storyboard_md5,
+            # yyMMddHHmmss
+        ).encode(),
+    ).hexdigest()
+
+    score_data = [
+        beatmap_md5,
+        username + (" " if has_supporter else ""),
+        score_online_checksum,
+        str(n300),
+        str(n100),
+        str(n50),
+        str(ngeki),
+        str(nkatu),
+        str(nmiss),
+        str(score),
+        str(max_combo),
+        str(perfect),
+        str(grade),
+        str(mods),
+        str(passed),
+        str(game_mode),
+        client_time.strftime("%y%m%d%H%M%S"),
+        str(osu_version),
+        "26685362",  # TODO what is this?
+    ]
+
+    iv_b64 = b"N2Q1YWZiNzYzNWFiYWZjZWMyMWMwM2QwMDEzOGRiNDk="
+    visual_settings_b64 = b"YHD/rr/lajZIr+ZC6UbFYvCOwTOaEF3qhJCFaZUlQA8="
+
+    score_data_b64, client_hash_b64 = encryption.encrypt_score_aes_data(
+        score_data,
+        client_hashes,
+        iv_b64=iv_b64,
+        osu_version=osu_version,
+    )
+    score_time = 13358
+    fail_time = 0
+    exited_out = False
+
+    # ACT
+    response = await http_client.post(
+        url="/web/osu-submit-modular-selector.php",
+        headers={"Host": "osu.cmyui.xyz", "token": "auth-token"},
+        data={
+            "x": exited_out,
+            "ft": fail_time,
+            "fs": visual_settings_b64,
+            "bmk": beatmap_md5,  # (`updated_beatmap_hash` in code)
+            "sbk": storyboard_md5,
+            "iv": iv_b64,
+            "c1": f"{uninstall_md5}|{disk_signature_md5}",
+            "st": score_time,
+            "pass": passwd_md5,
+            "osuver": osu_version,
+            "s": client_hash_b64,
+            # score param
+            "score": score_data_b64,
+        },
+        files={
+            # simulate replay data
+            "score": b"12345"
+            * 100,
+        },
+    )
+
+    # ASSERT
+    assert response.status_code == status.HTTP_200_OK
```

Comment:
we should actually check the response doesn't start with "error: " here, because we respond with 200 even on error cases

Reply:
yes i still have a lot of things to assert - would be perhaps good to install some function call spies as well to measure what's being performed

Reply:
yep, we can probably assert the entire chart response even

Reply:
i will try to get this running in ci & merged so we can iterate on it in future prs